include/ginkgo/core/base/scoped_device_id.hpp

```diff
@@ -0,0 +1,228 @@
+
+    hip_scoped_device_id(hip_scoped_device_id&& other) noexcept;
+
+    hip_scoped_device_id& operator=(hip_scoped_device_id&& other) noexcept;
+
+private:
+    int original_device_id_;
+    bool need_reset_;
+};
+
+
+}  // namespace detail
+
+
+/**
+ * This move-only class uses RAII to set the device id within a scoped block, if
+ * necessary.
+ *
+ * The class behaves similar to std::scoped_lock. The scoped guard will make
+ * sure that the device code is run on the correct device within one scoped
+ * block, when run with multiple devices. Depending on the executor it will
+ * record the current device id and set the device id to the one being passed
+ * in. After the scope has been exited, the destructor sets the device_id back
+ * to the one before entering the scope. The OmpExecutor and DpcppExecutor don't
+ * require setting an device id, so in those cases, the class is a no-op.
+ *
+ * The device id scope has to be constructed from a executor with concrete type
+ * (not plain Executor) and a device id. Only the type of the executor object is
+ * relevant, so the pointer will not be accessed, and may even be a nullptr.
+ * From the executor type the correct derived class of
+ * detail::generic_scoped_device_id is picked. The following illustrates the
+ * usage of this class:
+ * ```
+ * {
+ *   scoped_device_id g{static_cast<CudaExecutor>(nullptr), 1};
+ *   // now the device id is set to 1
+ * }
+ * // now the device id is reverted again
+ * ```
+ */
+class scoped_device_id {
+public:
+    /**
+     * Create a scoped device id from an OmpExecutor.
+     *
+     * This will pick the noop_scoped_device_id.
+     *
+     * @param exec  Not used.
+     * @param device_id  Not used.
+     */
+    scoped_device_id(const OmpExecutor* exec, int device_id)
+        : scope_(std::make_unique<detail::noop_scoped_device_id>())
+    {}
+
+    /**
+     * Create a scoped device id from an CudaExecutor.
+     *
+     * This will pick the cuda_scoped_device_id.
+     *
+     * @param exec  Not used.
+     * @param device_id  The device id to use within the scope.
+     */
+    scoped_device_id(const CudaExecutor* exec, int device_id)
+        : scope_(std::make_unique<detail::cuda_scoped_device_id>(device_id))
+    {}
+
+    /**
+     * Create a scoped device id from an HipExecutor.
+     *
+     * This will pick the hip_scoped_device_id.
+     *
+     * @param exec  Not used.
+     * @param device_id  The device id to use within the scope.
+     */
+    scoped_device_id(const HipExecutor* exec, int device_id)
+        : scope_(std::make_unique<detail::hip_scoped_device_id>(device_id))
+    {}
+
+    /**
+     * Create a scoped device id from an OmpExecutor.
+     *
+     * This will pick the noop_scoped_device_id.
+     *
+     * @param exec  Not used.
+     * @param device_id  Not used.
+     */
+    scoped_device_id(const DpcppExecutor* exec, int device_id)
+        : scope_(std::make_unique<detail::noop_scoped_device_id>())
+    {}
```

Comment:
implementations don't need to live in the header

Reply:
I think this doesn't work. I've put the implementation of `EXEC::get_scoped_device_id` into the `ginkgo_*_device` library, so these need the constructor to be available. But since these libraries are not linked against core, this results in linker errors.
I think we have two options:
1. `scoped_device_id` in core, `EXEC::get_scoped_device_id` in header
2. `scoped_device_id` in header, `EXEC::get_scoped_device_id` in devices
The easiest would be 2., because then I would just drop the last commit. Not sure if there is an actual benefit to one of them.